crates/walrus-indexer/src/storage.rs

```diff
@@ -0,0 +1,471 @@
+    upper_bound.extend_from_slice(patch_blob_id.as_ref());
+    upper_bound.push(0x01);
+
+    (lower_bound, upper_bound)
+}
+
+// Parse the quilt patch index key.
+fn parse_quilt_patch_index_key(key: &[u8]) -> Result<(BlobId, ObjectID), anyhow::Error> {
+    let key_struct = QuiltPatchIndexKey::from_key(key)?;
+    Ok((key_struct.patch_blob_id, key_struct.object_id))
+}
+
+// A concise representation of a quilt patch.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Patch {
+    /// The identifier.
+    pub identifier: String,
+    /// The internal ID, can be combined with quilt_id to get QuiltPatchId.
+    pub quilt_patch_internal_id: Vec<u8>,
+    /// The patch blob ID.
+    pub patch_blob_id: Option<BlobId>,
+}
+
+/// Extracts Patches from a QuiltIndex.
+fn get_patches_for_primary_index(quilt_index: &QuiltIndex) -> Vec<Patch> {
+    assert!(matches!(quilt_index, QuiltIndex::V1(_)));
+    let QuiltIndex::V1(quilt_index_v1) = quilt_index;
+    quilt_index_v1
+        .patches()
+        .iter()
+        .map(|patch| Patch {
+            identifier: patch.identifier.clone(),
+            quilt_patch_internal_id: patch.quilt_patch_internal_id().to_bytes(),
+            // TODO: Add the actual blob id when this is supported.
+            patch_blob_id: None,
+        })
+        .collect()
+}
+
+/// The target of an index entry, it could be a blob, a quilt patch or a quilt.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum IndexTarget {
+    /// Blob ID.
+    Blob(BlobIdentity),
+    /// Quilt patch ID.
+    QuiltPatchId(QuiltPatchId),
+}
+
+/// Blob identity containing both blob_id and object_id.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct BlobIdentity {
+    /// The blob ID.
+    pub blob_id: BlobId,
+    /// The Sui object ID.
+    pub object_id: ObjectID,
+}
+
+/// Value stored in the object index.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ObjectIndexValue {
+    /// The blob manager ID.
+    pub blob_manager: ObjectID,
+    /// The identifier.
+    pub identifier: String,
+    /// The sequence number.
+    pub sequence_number: u64,
+}
+
+/// Storage interface for the Walrus Index (Dual Index System).
+#[derive(Debug, Clone)]
+pub struct WalrusIndexStore {
+    // Reference to the underlying RocksDB instance for transaction support.
+    db: Arc<typed_store::rocks::RocksDB>,
+    // Primary index: blob_manager/identifier -> IndexTarget.
+    primary_index: DBMap<String, IndexTarget>,
+    // Object index: object_id -> blob_manager/identifier.
+    object_index: DBMap<String, ObjectIndexValue>,
+    // Sequence store: stores the last processed sequence number for resumption.
+    sequence_store: DBMap<String, u64>,
+    // Index for quilt patches, so that we can look up quilt patches by the corresponding
+    // files' blob IDs.
+    quilt_patch_index: DBMap<String, QuiltPatchId>,
```

Comment:
A couple of questions:
- primary_index: makes a lot of sense! (not really a question)
- what is object_index will be used for? Lifetime management?
- Unclear what is `sequence`
- for quilt_patch_index, based on the comment, the key is `files' blob IDs`, but the key of the DBMap is String. Is it supposed to be BlobID or the comment is out dated?

Reply:
I see. I think I didn't connect the String with the to_key() function earlier. SGTM